src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1198 @@
+}
+
+size_t VST3Effect::ProcessBlock(float** inBlock, float** outBlock, size_t blockLen)
+{
+   internal::ComponentHandler::PendingChangesPtr pendingChanges { nullptr };
+   if(mComponentHandler)
+      pendingChanges = mComponentHandler->getPendingChanges();
+   return VST3ProcessBlock(mEffectComponent.get(), mSetup, inBlock, outBlock, blockLen, pendingChanges.get());
+}
+
+bool VST3Effect::RealtimeInitialize()
+{
+   //reload current parameters form the editor into parameter queues
+   SyncParameters();
+   return true;
+}
+
+bool VST3Effect::RealtimeAddProcessor(unsigned numChannels, float sampleRate)
+{
+   using namespace Steinberg;
+
+   try
+   {
+      auto effect = std::make_unique<VST3Effect>(*this);
+      effect->mSetup.processMode = Vst::kRealtime;
+      effect->mSetup.sampleRate = sampleRate;
+      if(!effect->ProcessInitialize({0}, nullptr))
+         throw std::runtime_error { "VST3 realtime initialization failed" };
+
+      mRealtimeGroupProcessors.push_back(std::move(effect));
+      return true;
+   }
+   catch(std::exception& e)
+   {
+      //make_unique<> isn't likely to fail here since this effect instance
+      //somehow succeeded
+      wxLogError("Failed to add realtime processor: %s", e.what());
+   }
+   return false;
+}
+
+bool VST3Effect::RealtimeFinalize()
+{
+   for(auto& processor : mRealtimeGroupProcessors)
+      processor->ProcessFinalize();
+
+   mRealtimeGroupProcessors.clear();
+   mPendingChanges.reset();
+   
+   return true;
+}
+
+bool VST3Effect::RealtimeSuspend()
+{
+   for(auto& effect : mRealtimeGroupProcessors)
+      effect->mAudioProcessor->setProcessing(false);
+   return true;
+}
+
+bool VST3Effect::RealtimeResume() noexcept
+{
+   for(auto& effect : mRealtimeGroupProcessors)
+      effect->mAudioProcessor->setProcessing(true);
+   return true;
+}
+
+bool VST3Effect::RealtimeProcessStart()
+{
+   assert(mPendingChanges == nullptr);
+
+   if(mComponentHandler != nullptr)
+      //Same parameter changes are used among all of the relatime processors
+      mPendingChanges = mComponentHandler->getPendingChanges();
+   return true;
+}
+
+size_t VST3Effect::RealtimeProcess(int group, float** inBuf, float** outBuf, size_t numSamples)
+{
+   auto& effect = mRealtimeGroupProcessors[group];
+   return VST3ProcessBlock(effect->mEffectComponent.get(), effect->mSetup, inBuf, outBuf, numSamples, mPendingChanges.get());
+}
+
+bool VST3Effect::RealtimeProcessEnd() noexcept
+{
+   mPendingChanges.reset();
+   return true;
+}
+
+int VST3Effect::ShowClientInterface(wxWindow& parent, wxDialog& dialog, bool forceModal)
+{
+   if(!IsGraphicalUI())
+   {
+      //Restrict resize of the "plain" dialog
+      dialog.SetMaxSize(dialog.GetSize());
+      dialog.SetMinSize(dialog.GetSize());
+   }
+   if(forceModal)
+      return dialog.ShowModal();
+
+   dialog.Show();
+   return 0;
+}
+
+bool VST3Effect::SetHost(EffectHostInterface* host)
+{
+   mEffectHost = host;
+
+   if(host)
+   {
+      ReloadUserOptions();
```

Comment:
`SetHost` is now no longer empty, but it is still incomplete.  See how the Private configuration "Parameters" is loaded and stored for the other effect families.